### ISSUE TITLE
project's "processes" tab does not load

### ISSUE THREAD

**Issue by haraldschilly** [sagemathinc/cocalc#5081]
sometimes the "processes" tab is stuck trying to load any data, i.e. just `O Loading...`and that's it.

workaround: restart the project

**Comment by haraldschilly** [comment]
I tried to gather some log data. Here is a block, which ends in "closed"

```
2020-12-10T18:02:18.626Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  send_versioned_changes_to_browsers
2020-12-10T18:02:20.631Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  send_versioned_changes_to_browsers
2020-12-10T18:02:22.638Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  send_versioned_changes_to_browsers
2020-12-10T18:02:24.644Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  send_versioned_changes_to_browsers
2020-12-10T18:02:26.651Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  send_versioned_changes_to_browsers
2020-12-10T18:02:28.657Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  send_versioned_changes_to_browsers
2020-12-10T18:02:30.666Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  send_versioned_changes_to_browsers
2020-12-10T18:02:32.215Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  check_if_should_close --  0  -- do a save and maybe close
2020-12-10T18:02:32.215Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  save_and_close_if_possible: no connections, so saving...
2020-12-10T18:02:32.215Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  save_and_close_if_possible: after save there are 0 connections and delay=61001
2020-12-10T18:02:32.216Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  save_and_close_if_possible: close this SyncTableChannel atomically
2020-12-10T18:02:32.216Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  close: closing
2020-12-10T18:02:32.217Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  close: closing
2020-12-10T18:02:32.217Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  close: closed
2020-12-10T18:02:32.217Z - debug: project_info close
```

and a bit later, when accessing it again:

```
2020-12-10T18:06:01.627Z - debug: primus-api request object {"cmd":"project_info"}                   
2020-12-10T18:06:01.881Z - debug: project_info: channel: new connection from 77.119.130.241 -- 2RLe2Chma_ylLdTWgIItq:0yt1b37p6y9:project_info
2020-12-10T18:06:01.887Z - debug: primus-api request object {"cmd":"synctable_channel","query":{"project_info":[{"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164","info":null}]},"options":[{"ephemeral":true}]}
2020-12-10T18:06:01.888Z - debug: synctable_channel {"project_info":[{"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164","info":null}]} sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813
2020-12-10T18:06:01.889Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  creating new sync channel (persistent=false, ephemeral=true)
2020-12-10T18:06:01.889Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  init_handlers
2020-12-10T18:06:01.890Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  init_synctable
2020-12-10T18:06:01.890Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  init_synctable -- ephemeral (no database)
2020-12-10T18:06:01.892Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  created synctable -- waiting for connected state
2020-12-10T18:06:01.895Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  created synctable -- now connected
2020-12-10T18:06:01.895Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  broadcast_synctable_to_browsers
2020-12-10T18:06:01.895Z - debug: register_project_info_table                                                                   
2020-12-10T18:06:01.895Z - debug: register_project_info_table: cleaning up an already existing one
2020-12-10T18:06:02.310Z - debug: primus-api request object {"cmd":"synctable_channel","query":{"project_info":[{"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164","info":null}]},"options":[{"ephemeral":true}]}
2020-12-10T18:06:02.310Z - debug: synctable_channel {"project_info":[{"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164","info":null}]} sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813                 
2020-12-10T18:06:02.545Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  new connection from (address=77.119.130.241, conn=2RLe2Chma_ylLdTWgIItq) -- 2RLe2C
hma_ylLdTWgIItq:revsq9fenun:sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813 -- num_connections = 1 (from this client = 1)                                                                    
2020-12-10T18:06:02.545Z - debug: SyncTableChannel('sync:b49bfaf156c5f4d2ff594470bdd160b84f8b2813', '{"project_info":[{"info":null,"project_id":"fe935bd1-ffba-46a1-8602-d1fcca575164"}]}'):  send_synctable_to_browser

```

**Comment by williamstein** [comment]
I recommend changing parameters so the project info table *never* ever closes.  Once it is created, just leave it there.  It doesn't use a lot of memory and there are many other tables (e.g. for jupyter notebooks, the directory listings table, etc.) that persist until very explicitly closed.

**Comment by williamstein** [comment]
Also, showing the info tab again for a user will be faster if the table doesn't have to be created again, which is a benefit (it's annoying that there is a nontrivial pause when you click on the info tab).

**Comment by haraldschilly** [comment]
Sorry, I don't know how to do such a change. This is copy/paste of code where I don't know what it is really doing.

And well, while the tab isn't shown, no data updates should happen. It's sending messages all the time.

**Comment by haraldschilly** [comment]
Today I saw the same happening with the project status. There was a lot of cpu usage but no cpu usage alert popping up. I refreshed cocalc in the browser tab, and the cpu alert showed up immediately. However. a few minutes later (still cpu usage high), the alert was gone. Also, the project's processes tab no longer loads.

**Comment by williamstein** [comment]
Maybe you should put in a quick note that just says "if nothing appears here, please restart your project" with a link to this page?

**Comment by williamstein** [comment]
I noticed that in an example project where saving an ipynb file to disk is giving an error (not for the ipynb file but for an auxiliary temp file -- so no loss of data!), this processes info is ALSO not loading.  So I suspect there may be a connection between the two bugs; maybe they are the same bug!  Perhaps something about how the process info is implemented is causing this other bug?).     On that project, I've been watching the websocket traffic, and there is an actual error:

![image](https://user-images.githubusercontent.com/1276278/110853527-969ccc00-8268-11eb-9452-b545cd5e9858.png)

My guess is that there is some cleanup code in the process info implementation, which is calling off on something undefined.  We'll see...  In any case, this is a clue.

**Comment by williamstein** [comment]
OK, I found two dangerous calls to off in that code, and fixed them in f77220181dfaf496cb0fe060a8f9ce8c8dc9ee99

**Comment by williamstein** [comment]
I'm closing this, since I fixed something that would potentially cause this bug, and also I can't reproduce this.  Re-open if this happens on a project that was restarted *after* I left this message.

**Comment by haraldschilly** [comment]
ok, nice, I'll keep my eyes open to see if happens again

**Comment by cmielke-vir** [comment]
Happening to me - and points for calling this one a Heisenbug 😛 

For me, this happens when I do long-running jobs overnight with a machine-learning hyperparameter tuner (keras-tune), which runs some heavy-gpu processes and prints continuing TQDM progress bars into the cell. I wonder if that continual output could cause the issue?
